File: registry-compiler/src/main/java/com/github/fengdai/registry/compiler/RegistryClass.java
package com.github.fengdai.registry.compiler;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.FieldSpec;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.ParameterSpec;
import com.squareup.javapoet.ParameterizedTypeName;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.TypeSpec;
import com.squareup.javapoet.WildcardTypeName;
import java.util.ArrayList;
import java.util.Collection;
import javax.lang.model.element.Element;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;

import static com.google.auto.common.MoreElements.getPackage;

final class RegistryClass {
  private static final String REGISTRY_IMPL_SUFFIX = "Registry";
  private static final String ITEM_IMPL_NAME = "Item";

  private static final ClassName ILLEGAL_ARGUMENT_EXCEPTION =
      ClassName.get(IllegalArgumentException.class);
  private static final ClassName ANDROID_VIEW_HOLDER =
      ClassName.get("android.support.v7.widget", "RecyclerView", "ViewHolder");
  private static final ParameterizedTypeName ANDROID_VIEW_HOLDER_CLASS =
      ParameterizedTypeName.get(ClassName.get(Class.class),
          WildcardTypeName.subtypeOf(ANDROID_VIEW_HOLDER));

  private static final ClassName REGISTRY =
      ClassName.get("com.github.fengdai.registry", "Registry");
  private static final ClassName ITEM =
      ClassName.get("com.github.fengdai.registry", "Registry", "Item");
  private static final ClassName BINDER =
      ClassName.get("com.github.fengdai.registry", "Binder");

  private final Element originatingElement;
  private final ClassName annotation;
  private final String classPackage;
  private final String className;
  private final Collection<MappingInfo> mappingInfoList;
  private final Collection<IndexedViewHolderInfo> indexedViewHolders;
  private final Collection<StaticContentLayout> staticContentLayouts;

  RegistryClass(TypeElement annotationElement, Collection<MappingInfo> mappingInfoList,
      Collection<IndexedViewHolderInfo> indexedViewHolders,
      Collection<Id> staticContentLayouts) {
    this.originatingElement = annotationElement;
    this.annotation = ClassName.get(annotationElement);
    this.classPackage = getPackage(annotationElement).getQualifiedName().toString();
    this.className = annotationElement.getQualifiedName()
        .toString()
        .substring(classPackage.length() + 1)
        .replace('.', '$') + REGISTRY_IMPL_SUFFIX;
    this.mappingInfoList = mappingInfoList;
    this.indexedViewHolders = indexedViewHolders;

    int staticContentLayoutIndex = indexedViewHolders.size();
    this.staticContentLayouts = new ArrayList<>(staticContentLayouts.size());
    for (Id layout : staticContentLayouts) {
      this.staticContentLayouts.add(new StaticContentLayout(staticContentLayoutIndex, layout));
      staticContentLayoutIndex++;
    }
  }

  JavaFile brewJava() {
    return JavaFile.builder(classPackage, buildRegistryImpl())
        .addFileComment("Generated code from Registry. Do not modify!")
        .build();
  }

  private TypeSpec buildRegistryImpl() {
    TypeSpec.Builder builder = TypeSpec.classBuilder(className)
        .addOriginatingElement(originatingElement)
        .addAnnotation(annotation)
        .addModifiers(Modifier.PUBLIC, Modifier.FINAL);

    ClassName itemClassName = ClassName.get(classPackage, className, ITEM_IMPL_NAME);
    builder.addType(itemClass(itemClassName));
    builder.superclass(ParameterizedTypeName.get(REGISTRY, itemClassName));

    for (MappingInfo mappingInfo : mappingInfoList) {
      builder.addMethod(itemOfMethod(itemClassName, mappingInfo));
    }

    for (StaticContentLayout layout : staticContentLayouts) {
      FieldSpec staticContentLayoutItemField =
          FieldSpec.builder(itemClassName, "ITEM_" + layout.id.resourceName)
              .addModifiers(Modifier.PRIVATE, Modifier.FINAL, Modifier.STATIC)
              .initializer("new $T(staticContentLayoutData($L), $L, null)", itemClassName,
                  layout.id.code,
                  layout.viewType)
              .build();
      MethodSpec staticContentLayoutItemGetter =
          MethodSpec.methodBuilder("itemOf_" + layout.id.resourceName)
              .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
              .returns(itemClassName)
              .addStatement("return $N", staticContentLayoutItemField)
              .build();
      builder.addField(staticContentLayoutItemField);
      builder.addMethod(staticContentLayoutItemGetter);
    }

    // Constructor
    builder.addMethod(constructor(indexedViewHolders, staticContentLayouts));
    return builder.build();
  }

  private MethodSpec itemOfMethod(ClassName itemClassName, MappingInfo mappingInfo) {
    MethodSpec.Builder builder = MethodSpec.methodBuilder("itemOf")
        .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
        .addParameter(ParameterSpec.builder(mappingInfo.getDataClassName(), "data").build())
        .returns(itemClassName);
    if (!mappingInfo.isOneToMany()) {
      MappingInfo.Mapping mapping = mappingInfo.getMappings().get(0);
      if (mapping.binderInfo == null) {
        builder.addStatement("return new $T(data, $L, BINDABLE_VIEW_HOLDER_BINDER)",
            itemClassName, mapping.viewType);
      } else {
        builder.addStatement("return new $T(data, $L, new $T())",
            itemClassName, mapping.viewType, mapping.binderInfo.binderClassName);
      }
    } else {
      builder.addParameter(
          ParameterSpec.builder(ANDROID_VIEW_HOLDER_CLASS, "viewHolderClass").build());

      int i = 0;
      for (MappingInfo.Mapping mapping : mappingInfo.getMappings()) {
        if (i == 0) {
          builder.beginControlFlow("if (viewHolderClass == $T.class)",
              mapping.viewHolderInfo.viewHolderClassName);
        } else {
          builder.nextControlFlow("else if (viewHolderClass == $T.class)",
              mapping.viewHolderInfo.viewHolderClassName);
        }
        if (mapping.binderInfo == null) {
          builder.addStatement("return new $T(data, $L, BINDABLE_VIEW_HOLDER_BINDER)",
              itemClassName, mapping.viewType);
        } else {
          builder.addStatement("return new $T(data, $L, new $T())",
              itemClassName, mapping.viewType, mapping.binderInfo.binderClassName);
        }
        i++;
      }
      builder.nextControlFlow("else")
          .addStatement("throw new $T($S)", ILLEGAL_ARGUMENT_EXCEPTION, "viewHolderClass")
          .endControlFlow();
    }
    return builder.build();
  }

  private static MethodSpec constructor(Collection<IndexedViewHolderInfo> indexedViewHolders,
      Collection<StaticContentLayout> staticContentLayouts) {
    MethodSpec.Builder builder = MethodSpec.constructorBuilder()
        .addModifiers(Modifier.PUBLIC)
        .addStatement("super()");

    // registerViewHolderFactory
    for (IndexedViewHolderInfo viewHolder : indexedViewHolders) {
      ParameterSpec factoryParameter = null;
      if (!viewHolder.info.unresolvedDependencies.isEmpty()) {
        factoryParameter = ParameterSpec.builder(viewHolder.info.factoryImplClassName,
            "factory" + viewHolder.index).build();
        builder.addParameter(factoryParameter);
      }
      builder.addStatement(registerViewHolderFactory(viewHolder, factoryParameter));
    }

    // registerStaticContentLayout
    for (StaticContentLayout layout : staticContentLayouts) {
      builder.addStatement(registerStaticContentLayout(layout));
    }

    return builder.build();
  }

  private static CodeBlock registerViewHolderFactory(IndexedViewHolderInfo viewHolder,
      ParameterSpec factoryParameter) {
    CodeBlock factory = factoryParameter != null
        ? CodeBlock.of("$N", factoryParameter)
        : CodeBlock.of("new $T()", viewHolder.info.factoryImplClassName);
    return CodeBlock.of("registerViewHolderFactory($L, $L)", viewHolder.index, factory);
  }

  private static CodeBlock registerStaticContentLayout(StaticContentLayout layout) {
    return CodeBlock.of("registerStaticContentLayout($L, $L)", layout.viewType, layout.id.code);
  }

  private static TypeSpec itemClass(ClassName name) {
    return TypeSpec.classBuilder(name)
        .superclass(ITEM)
        .addModifiers(Modifier.PUBLIC, Modifier.FINAL, Modifier.STATIC)
        .addMethod(MethodSpec.constructorBuilder()
            .addModifiers(Modifier.PRIVATE)
            .addParameter(ParameterSpec.builder(TypeName.OBJECT, "data").build())
            .addParameter(ParameterSpec.builder(TypeName.INT, "viewType").build())
            .addParameter(ParameterSpec.builder(BINDER, "binder").build())
            .addStatement("super(data, viewType, binder)")
            .build())
        .build();
  }
}

File: registry-processor/src/main/java/com/github/fengdai/registry/processor/RegistryClass.kt
package com.github.fengdai.registry.processor

import androidx.annotation.NonNull
import androidx.annotation.Nullable
import com.github.fengdai.registry.processor.internal.applyEach
import com.github.fengdai.registry.processor.internal.peerClassWithReflectionNesting
import com.squareup.javapoet.ClassName
import com.squareup.javapoet.CodeBlock
import com.squareup.javapoet.FieldSpec
import com.squareup.javapoet.MethodSpec
import com.squareup.javapoet.ParameterSpec
import com.squareup.javapoet.ParameterizedTypeName
import com.squareup.javapoet.TypeName
import com.squareup.javapoet.TypeSpec
import javax.lang.model.element.Modifier.FINAL
import javax.lang.model.element.Modifier.PRIVATE
import javax.lang.model.element.Modifier.PUBLIC
import javax.lang.model.element.Modifier.STATIC

private val BINDER = ClassName.get("com.github.fengdai.registry", "Binder")
private val ADAPTER_DELEGATE = ClassName.get("com.github.fengdai.registry", "AdapterDelegate")
private val VIEW_HOLDER_FACTORY = ClassName.get("com.github.fengdai.viewholder", "ViewHolderFactory")
private val VIEW_GROUP = ClassName.get("android.view", "ViewGroup")
private val SPARSE_ARRAY = ClassName.get("android.util", "SparseArray")
private val VIEW_HOLDER = ClassName.get("androidx.recyclerview.widget", "RecyclerView", "ViewHolder")
private val LAYOUT_VIEW_HOLDER_FACTORY = ClassName.get("com.github.fengdai.registry.internal", "LayoutViewHolderFactory")
private val BINDERS = ClassName.get("com.github.fengdai.registry.internal", "Binders")
private val DATA_RESOLVER = ClassName.get("com.github.fengdai.registry.internal", "DataResolver")

data class RegistryClass(
  val targetType: ClassName,
  val itemType: ClassName,
  val public: Boolean,
  val injected: Boolean,
  val bindingSets: List<BindingSet>,
  val indexedViewHolderTypes: List<Pair<Int, TypeName>>,
  val layouts: List<LayoutBinding>
) {
  val generatedType: ClassName = targetType.run { peerClassWithReflectionNesting(simpleName() + "_Impl") }
  private val generatedItemType = generatedType.nestedClass("Item")
  private val generatedAdapterDelegateType = generatedType.nestedClass("AdapterDelegate")
  private val sortedIndexedViewHolders = indexedViewHolderTypes.sortedBy { it.second.toString() }
  private val indexedLayouts: List<Pair<Int, LayoutBinding>> = layouts
      .mapIndexed { index, id -> Pair(index + indexedViewHolderTypes.size, id) }

  fun brewJava(): TypeSpec {
    return TypeSpec.classBuilder(generatedType)
        .addModifiers(FINAL)
        .apply {
          if (public) {
            addModifiers(PUBLIC)
          }
        }
        .addSuperinterface(targetType)
        .applyEach(bindingSets) {
          it.bindings.forEach { binding ->
            val dataResolverField =
              FieldSpec.builder(DATA_RESOLVER, binding.dataResolverFieldName, PRIVATE, FINAL)
                  .initializer(CodeBlock.of(
                      "new \$T(\$L, \$L)", DATA_RESOLVER, binding.viewType,
                      if (binding.targetIsBinder) CodeBlock.of("new \$T()", binding.targetType)
                      else CodeBlock.of("\$T.BINDER_VIEW_HOLDER_BINDER", BINDERS)))
                  .build()
            addField(dataResolverField)
            addMethod(MethodSpec.overriding(binding.factoryMethod)
                .addStatement(CodeBlock.of(
                    "return new \$T(\$L, \$N)", generatedItemType,
                    binding.factoryMethod.parameters.single().simpleName,
                    dataResolverField))
                .build())
          }
        }
        .applyEach(indexedLayouts) {
          val (viewType, layoutBinding) = it
          val factoryMethod = layoutBinding.factoryMethod
          val layout = layoutBinding.layout
          val field =
            FieldSpec.builder(itemType, "item_" + layout.resourceName)
                .addModifiers(PRIVATE, FINAL)
                .initializer("new \$T(\$L, new \$T(\$L, null))", generatedItemType, layout.code, DATA_RESOLVER, viewType)
                .build()
          addField(field)
          addMethod(MethodSpec.overriding(factoryMethod)
              .addStatement("return \$N", field)
              .build())
        }
        .addType(TypeSpec.classBuilder(generatedItemType)
            .addSuperinterface(itemType)
            .addModifiers(STATIC, PRIVATE, FINAL)
            .addField(TypeName.OBJECT, "data", PRIVATE, FINAL)
            .addField(DATA_RESOLVER, "dataResolver", PRIVATE, FINAL)
            .addMethod(MethodSpec.constructorBuilder()
                .addParameter(TypeName.OBJECT, "data")
                .addParameter(DATA_RESOLVER, "dataResolver")
                .addStatement("this.data = data")
                .addStatement("this.dataResolver = dataResolver")
                .build())
            .addMethod(MethodSpec.methodBuilder("getData")
                .addModifiers(PUBLIC)
                .addAnnotation(Override::class.java)
                .returns(TypeName.OBJECT)
                .addStatement("return data")
                .build())
            .addMethod(MethodSpec.methodBuilder("getViewType")
                .addModifiers(PUBLIC)
                .addAnnotation(Override::class.java)
                .returns(TypeName.INT)
                .addStatement("return dataResolver.viewType")
                .build())
            .addMethod(MethodSpec.methodBuilder("getBinder")
                .addModifiers(PUBLIC)
                .addAnnotation(Nullable::class.java)
                .addAnnotation(Override::class.java)
                .returns(BINDER)
                .addStatement("return dataResolver.binder")
                .build())
            .build())
        .addType(TypeSpec.classBuilder(generatedAdapterDelegateType)
            .addModifiers(STATIC, FINAL)
            .apply {
              if (public) {
                addModifiers(PUBLIC)
              }
            }
            .superclass(ParameterizedTypeName.get(ADAPTER_DELEGATE, itemType))
            .addField(FieldSpec.builder(ParameterizedTypeName.get(SPARSE_ARRAY, VIEW_HOLDER_FACTORY), "viewHolderFactories", PRIVATE, FINAL)
                .initializer("new SparseArray<>()")
                .build())
            .addMethod(MethodSpec.constructorBuilder()
                .addModifiers(PUBLIC)
                .apply {
                  if (injected) {
                    addAnnotation(ClassName.get("javax.inject", "Inject"))
                  }
                }
                .applyEach(sortedIndexedViewHolders) {
                  addParameter(
                      ParameterizedTypeName.get(VIEW_HOLDER_FACTORY, it.second), "factory${it.first}"
                  )
                }
                .addStatement("super()")
                .applyEach(indexedViewHolderTypes) {
                  addStatement("viewHolderFactories.put(\$L, \$L)", it.first, "factory${it.first}")
                }
                .applyEach(indexedLayouts) {
                  addStatement("viewHolderFactories.put(\$L, new \$T(\$L))", it.first, LAYOUT_VIEW_HOLDER_FACTORY, it.second.layout.code)
                }
                .build())
            .addMethod(MethodSpec.methodBuilder("onCreateViewHolder")
                .addAnnotation(Override::class.java)
                .addAnnotation(NonNull::class.java)
                .addModifiers(PUBLIC)
                .returns(VIEW_HOLDER)
                .addParameter(ParameterSpec.builder(VIEW_GROUP, "parent").build())
                .addParameter(ParameterSpec.builder(TypeName.INT, "viewType").build())
                .addStatement("return viewHolderFactories.get(viewType).create(parent)")
                .build())
            .build())
        .build()
  }
}

private val Binding.dataResolverFieldName get() = "resolver_" + factoryMethod.simpleName + "_" + dataRawType.reflectionName().replace('.', '_')
